compiler/rustc_hir_analysis/src/impl_wf_check.rs

```diff
@@ -235,8 +246,130 @@ pub(crate) fn enforce_impl_non_lifetime_params_are_constrained(
+        if let Some(res_def_id) = path.res.opt_def_id() {
+            if res_def_id == self.param_def_id {
+                self.found = true;
+                return ControlFlow::Break(()); // Found it, stop visiting.
+            }
+        }
+        // If not found, continue walking down the HIR tree.
+        intravisit::walk_path(self, path)
+    }
+
+    fn visit_lifetime(&mut self, lifetime: &'tcx rustc_hir::Lifetime) -> Self::Result {
+        if let LifetimeKind::Param(id) = lifetime.kind {
+            if let Some(local_def_id) = self.param_def_id.as_local() {
+                if id == local_def_id {
+                    self.found = true;
+                    return ControlFlow::Break(()); // Found it, stop visiting.
+                }
+            }
+        }
+        walk_lifetime(self, lifetime)
+    }
+}
+
+fn suggest_to_remove_or_use_generic(
+    tcx: TyCtxt<'_>,
+    diag: &mut Diag<'_>,
+    impl_def_id: LocalDefId,
+    param: &GenericParamDef,
+    parameter_type: &str,
+) {
+    let node = tcx.hir_node_by_def_id(impl_def_id);
+    let hir_impl = node.expect_item().expect_impl();
+
+    let Some((index, _)) = hir_impl
+        .generics
+        .params
+        .iter()
+        .enumerate()
+        .find(|(_, par)| par.def_id.to_def_id() == param.def_id)
+    else {
+        return;
+    };
+
+    // search if the parameter is used in the impl body
+    let mut visitor = ParamUsageVisitor {
+        tcx, // Pass the TyCtxt
+        param_def_id: param.def_id,
+        found: false,
+    };
+
+    for item_ref in hir_impl.items {
+        let _ = visitor.visit_impl_item_ref(item_ref);
+        if visitor.found {
+            break;
+        }
+    }
+
+    let is_param_used = visitor.found;
+
+    // Suggestion for removing the type parameter.
+    let mut suggestions = vec![];
+    if !is_param_used {
+        // Only suggest removing it if it's not used anywhere.
+        suggestions.push(vec![(hir_impl.generics.span_for_param_removal(index), String::new())]);
+    }
+
+    // Suggestion for making use of the type parameter.
+    if let Some(path) = extract_ty_as_path(hir_impl.self_ty) {
+        let seg = path.segments.last().unwrap();
+        if let Some(args) = seg.args {
+            suggestions
+                .push(vec![(args.span().unwrap().shrink_to_hi(), format!(", {}", param.name))]);
+        } else {
+            suggestions.push(vec![(seg.ident.span.shrink_to_hi(), format!("<{}>", param.name))]);
+        }
+    }
+
+    let msg = if is_param_used {
+        // If it's used, the only valid fix is to constrain it.
+        format!("make use of the {} parameter `{}` in the `self` type", parameter_type, param.name)
+    } else {
+        format!(
+            "either remove the unused {} parameter `{}`, or make use of it",
+            parameter_type, param.name
+        )
+    };
+
+    diag.multipart_suggestions(msg, suggestions, Applicability::MaybeIncorrect);
```

Comment:
I don't think we should use a multi-part suggestion for this, these should be multiple distinct suggestions